### PR TITLE
Update php language features to php 7.2

### PR DESCRIPTION
**Pull request by Tobion**
Using rector. Need to rebase after #10417

### CODE REVIEW

File: src/Composer/Command/DiagnoseCommand.php [hunk 1]
```diff
@@ -529,7 +529,7 @@ private function checkPlatform()
             $errors['json'] = true;
         }
 
-        if (!extension_loaded('Phar')) {
+        if (!extension_loaded(\Phar::class)) {
```
Comment (Tobion) [inline src/Composer/Command/DiagnoseCommand.php:532]: wrong change by rector

**Comment by Seldaek** [comment]
Needs a rebase again.. but TBH it's a lot of changes to review, and I've seen rector do bad mistakes in the past so the fact it's automated does not calm down that gut feeling at all.

As I wrote in https://github.com/composer/composer/pull/10421 - the ::class use seems useless to me if not done properly (as @GromNaN also commented) and same for https://github.com/composer/composer/pull/10419 and marking the consts public.

Overall looking through the changes I don't see much which is beneficial (the only one I saw is `isset() ternary` into `x ?? y` notation, that's a bit cleaner at least), and it carries lots of risk in review, so I'd rather close.

**Comment by Seldaek** [comment]
Closing as per the above comment, maybe we can do some automated code rewrites later but for now with the 2.2 branch needing maintenance I rather avoid too many changes unless they bring substantial gains to reduce odds of conflicts.

**Comment by Tobion** [comment]
> As I wrote in #10421 - the ::class use seems useless to me if not done properly (as @GromNaN also commented)

Totally don't agree with this. Semantically it is correct to use ::class. Adding use statements for the classes is just a matter of using a cs fixer. These are two separate things.

If you don't want constant visibility, I can just remove these parts. But I don't see why someone wouldn't want to use ::class

File: src/Composer/Command/ConfigCommand.php [hunk 2]
```diff
@@ -264,7 +264,7 @@ protected function execute(InputInterface $input, OutputInterface $output)
             $data = $this->config->all();
             if (Preg::isMatch('/^repos?(?:itories)?(?:\.(.+))?/', $settingKey, $matches)) {
                 if (!isset($matches[1]) || $matches[1] === '') {
-                    $value = isset($data['repositories']) ? $data['repositories'] : array();
+                    $value = $data['repositories'] ?? array();
```
[the remark this answers is not in the document]
Comment (Tobion) [reply src/Composer/Command/ConfigCommand.php:267]: unrelated as this is the job of a cs fixer

File: src/Composer/Package/Loader/RootPackageLoader.php [hunk 3]
```diff
@@ -69,9 +69,9 @@ public function __construct(RepositoryManager $manager, Config $config, VersionP
      *
      * @phpstan-param class-string<PackageClass> $class
      */
-    public function load(array $config, $class = 'Composer\Package\RootPackage', $cwd = null)
+    public function load(array $config, $class = \Composer\Package\RootPackage::class, $cwd = null)
     {
-        if ($class !== 'Composer\Package\RootPackage') {
+        if ($class !== \Composer\Package\RootPackage::class) {
```
[the remark this answers is not in the document]
Comment (Tobion) [reply src/Composer/Package/Loader/RootPackageLoader.php:74]: unrelated as this is the job of a cs fixer

File: src/Composer/Repository/VcsRepository.php [hunk 4]
```diff
@@ -78,21 +78,21 @@ public function __construct(array $repoConfig, IOInterface $io, Config $config,
...
-            'perforce' => 'Composer\Repository\Vcs\PerforceDriver',
-            'fossil' => 'Composer\Repository\Vcs\FossilDriver',
+            'github' => \Composer\Repository\Vcs\GitHubDriver::class,
+            'gitlab' => \Composer\Repository\Vcs\GitLabDriver::class,
+            'bitbucket' => \Composer\Repository\Vcs\GitBitbucketDriver::class,
+            'git-bitbucket' => \Composer\Repository\Vcs\GitBitbucketDriver::class,
+            'git' => \Composer\Repository\Vcs\GitDriver::class,
+            'hg' => \Composer\Repository\Vcs\HgDriver::class,
+            'perforce' => \Composer\Repository\Vcs\PerforceDriver::class,
+            'fossil' => \Composer\Repository\Vcs\FossilDriver::class,
```
[the remark this answers is not in the document]
Comment (Tobion) [reply src/Composer/Repository/VcsRepository.php:88]: unrelated as this is the job of a cs fixer

**Pull request closed without merging**